test/e2e/certificates_test.go

```diff
@@ -0,0 +1,120 @@
+		require.NoError(t, err, "error getting kubelet client CA ConfigMap")
+	}
+	// parse bundle from ConfigMap
+	bundleCABytes, err := certificates.GetCAsFromConfigMap(cm, certificates.CABundleKey)
+	if err != nil {
+		require.NoError(t, err, "error parsing CA bundle from ConfigMap")
+	}
+	bundleCAString := strings.TrimSpace(string(bundleCABytes))
+	// loop all Windows nodes
+	for _, winNode := range gc.allNodes() {
+		// spin-off subtests to ensure the CA bundle file is verified in all Windows nodes
+		t.Run("node/"+winNode.Name, func(t *testing.T) {
+			found, err := tc.isCertificateInCABundle(&winNode, bundleCAString)
+			require.NoError(t, err, "CA bundle file must be available")
+			assert.True(t, found, "certificate should be present in CA bundle")
+		})
+	}
+}
+
+// forceKubeApiServerCertificateRotation forces the certificate rotation by setting the `certificate-not-after`
+// annotation to nil in the `kube-apiserver-to-kubelet-signer` secret
+func (tc *testContext) forceKubeApiServerCertificateRotation() error {
+	// note that the slash had to be encoded as ~1 because it's a reference: https://www.rfc-editor.org/rfc/rfc6901#section-3
+	patchData := fmt.Sprint(`[{"op":"replace","path":"/metadata/annotations/auth.openshift.io~1certificate-not-after","value": null }]`)
+	// invoke request
+	_, err := tc.client.K8s.CoreV1().Secrets("openshift-kube-apiserver-operator").Patch(context.TODO(),
+		"kube-apiserver-to-kubelet-signer", types.JSONPatchType, []byte(patchData), meta.PatchOptions{})
+	return err
+}
+
+// isCertificateInCABundle returns True when the CA bundle file in the given Windows node contains the certificate
+// content, False otherwise. Waits retry.Interval, and run an SSH job to fetch the CA bundle, if the job fails,
+// retries every retry.Interval until retry.Timeout is reached.
+func (tc *testContext) isCertificateInCABundle(node *core.Node, content string) (bool, error) {
+	addr, err := controllers.GetAddress(node.Status.Addresses)
+	if err != nil {
+		return false, err
+	}
+	// CA bundle location in Windows node. i.e. "C:\k\kubelet-ca.crt"
+	caBundlePath := windows.GetK8sDir() + nodeconfig.KubeletClientCAFilename
+	// PowerShell command to fetch content in the file
+	command := fmt.Sprintf("Get-Content -Raw -Path %s", caBundlePath)
+	// wait retry.Interval and try to fetch the CA bundle
+	err = wait.Poll(retry.Interval, retry.Timeout, func() (bool, error) {
+		// invoke command
+		output, err := tc.runPowerShellSSHJob("kubelet-ca-bundle-content", command, addr)
+		if err != nil {
+			// retry
+			log.Printf("error fetching CA bundle in node %s with address %s, retrying in %s. %v",
+				node.Name, addr, retry.Interval, err)
+			return false, nil
+		}
+		// check certificate content in the CA bundle
+		if strings.Contains(output, content) {
+			// found
+			return true, nil
+		}
+		// certificate not found, invalid CA bundle
+		return false, errors.Errorf("invalid CA bundle in node %s with address %s", node.Name, addr)
```

Comment:
I would think you would want to retry here wouldn't you?
Its possible WMCO has not been able to copy the cert over yet for various reasons

Otherwise this function is only retrying in the case that the ca file wasnt fetchable, which it always should be, except for when there are network errors

Reply:
Agreed. Retry should be here and not in above case.

Reply:
Agreed ^ 2. Thanks for pointing this out. I refactored to `waitForKubeletCACertificateInNode(node *core.Node) error `. PTAL